spec/ob.dd

```diff
@@ -0,0 +1,309 @@
+
+$(HEADERNAV_TOC)
+
+$(B Experimental, Subject to Change)
+
+$(P If a memory object has only one pointer to it, that pointer is the $(I owner)
+of the memory object. With the single owner, it becomes straightforward to
+manage the memory for the object. It also becomes trivial to synchronize access
+to that memory object among multiple threads, because it can only be accessed
+by the thread that controls that single pointer.
+)
+
+$(P This can be generalized to a graph of memory objects interconnected by pointers,
+where only a single pointer connects to that graph from elsewhere. That single
+pointer becomes the $(I owner) of all the memory objects in that graph.
+)
+
+$(P When the owner of the graph is no longer needed, then the graph of memory
+objects it points to is no longer needed and can be safely disposed of. If the owner
+itself is no longer in use (i.e. is no longer $(I live)) and the owned memory
+objects are not disposed of, an error can be diagnosed.
+)
+
+$(P Hence, the following errors can be statically detected:
+)
+
+---
+int* allocate();    // allocate a memory object
+void release(int*); // deallocate a memory object
+
+@live void test()
+{
+    auto p = allocate();
+}   // error: p is not disposed of
+
+@live void test()
+{
+    auto p = allocate();
+    release(p);
+    release(p); // error: p was already disposed of
+}
+
+@live void test()
+{
+    auto p = void;
+    release(p); // error, p does not have a defined value
+}
+
+@live void test()
+{
+    auto p = allocate();
+    p = allocate(); // error: p was not disposed of
+    release(p);
+}
+---
+
+$(P Functions with the `@live` attribute enable diagnosing these sorts of errors by
+tracking the status of owner pointers.
+)
+
+$(H2 Ownership/Borrowing)
+
+$(P Tracking the ownership status of a pointer can be safely extended by adding the
+capability of temporarilly $(I borrowing) ownership of a pointer from the $(I owner).
+The $(owner) can no longer use the pointer as long as the $(I borrower) is still
+using the pointer value (i.e. is $(I live)). Once the $(I borrower) is no longer
+$(I live), the $(I owner) and resume using it. Only one $(I borrower) can be live
+at any point.
+)
+
+$(P Multiple $(I borrower) pointers can simultaneously exist if all of them are
+pointers to read only (`const` or `immutable`) data, i.e. none of them can modify the memory
+object(s) pointed to.
+)
+
+$(P This is collectively called an $(I Ownership/Borrowing) system. It can be state as:
+)
+
+$(BLOCKQUOTE At any point in the program, for each memory object,
+there is exactly one live mutable pointer to it or all the live pointers to
+it are read-only.)
+
+$(P Function declarations annotated with the `@live` attribute are checked for compliance with
+the Ownership/Borrowing rules. The checks are run after other semantic processing is complete.
+The checking does not influence code generation.
+)
+
+$(P Whether a pointer is allocates memory using the GC or some other
```

Comment:
Some grammatical error here with `is allocates`?

Reply:
fixed